File: common/src/main/java/com/ssblur/scriptor/helpers/targetable/Targetable.java
package com.ssblur.scriptor.helpers.targetable;

import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.core.Vec3i;
import net.minecraft.world.level.Level;
import net.minecraft.world.phys.Vec3;
import org.joml.Vector3f;

import javax.annotation.Nullable;

public class Targetable {
  Vec3 targetPos;
  BlockPos origin;
  Level level;
  Direction direction = null;
  Targetable finalTargetable;

  /**
   * A class describing anything that can be targeted by a spell
   * Shall always describe at least a position
   * @param pos The position targeted by / closest to the target of this cast
   */
  public Targetable(Level level, Vec3 pos) {
    targetPos = pos;
    this.level = level;
  }

  public Targetable(Level level, Vector3f pos) {
    targetPos = new Vec3(pos.x(), pos.y(), pos.z());
    this.level = level;
  }

  public Targetable(Level level, BlockPos pos) {
    targetPos = new Vec3(pos.getX(), pos.getY(), pos.getZ());
    this.level = level;
  }

  public Vec3 getTargetPos() {
    return targetPos;
  }
  public void setTargetPos(Vec3 targetPos) {
    this.targetPos = targetPos;
  }

  public void setTargetPos(Vector3f targetPos) {
    this.targetPos = new Vec3(targetPos.x(), targetPos.y(), targetPos.z());
  }

  public BlockPos getTargetBlockPos() {
    return new BlockPos((int) targetPos.x, (int) targetPos.y, (int) targetPos.z);
  }

  public Level getLevel() { return level; }

  public Direction getFacing() {
    if(direction != null)
      return direction;
    return Direction.UP;
  }

  public Targetable setFacing(@Nullable Direction direction) {
    this.direction = direction;
    return this;
  }

  public BlockPos getOffsetBlockPos() {
    if(direction != null)
      return getTargetBlockPos().relative(direction.getOpposite());
    return getTargetBlockPos();
  }

  public BlockPos getOrigin() {
    return origin;
  }

  public void setOrigin(BlockPos origin) {
    this.origin = origin;
  }

  public void setFinalTargetable(Targetable targetable) {
    this.finalTargetable = targetable;
  }

  public Targetable getFinalTargetable() {
    if(finalTargetable != null)
      return finalTargetable;
    return this;
  }

  public Targetable simpleCopy() {
    return new Targetable(level, targetPos);
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/helpers/targetable/Targetable.kt
package com.ssblur.scriptor.helpers.targetable

import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.world.level.Level
import net.minecraft.world.phys.Vec3
import org.joml.Vector3f
import kotlin.math.floor

@Suppress("unused")
open class Targetable {
  var targetPos: Vec3
  open val origin: BlockPos? = null
  var level: Level
  var direction: Direction? = null
  open var finalTargetable: Targetable? = this

  /**
   * A class describing anything that can be targeted by a spell
   * Shall always describe at least a position
   * @param pos The position targeted by / closest to the target of this cast
   */
  constructor(level: Level, pos: Vec3) {
    targetPos = pos
    this.level = level
  }

  constructor(level: Level, pos: Vector3f) {
    targetPos = Vec3(pos.x().toDouble(), pos.y().toDouble(), pos.z().toDouble())
    this.level = level
  }

  constructor(level: Level, pos: BlockPos) {
    targetPos = Vec3(pos.x.toDouble(), pos.y.toDouble(), pos.z.toDouble())
    this.level = level
  }

  fun setTargetPos(targetPos: Vector3f) {
    this.targetPos = Vec3(targetPos.x().toDouble(), targetPos.y().toDouble(), targetPos.z().toDouble())
  }

  val targetBlockPos: BlockPos
    get() = BlockPos(floor(targetPos.x).toInt(), floor(targetPos.y).toInt(), floor(targetPos.z).toInt())

  val facing: Direction
    get() {
      if (direction != null) return direction!!
      return Direction.UP
    }

  open val entityYRotation: Float? = null

  open val entityCoarseYRotation: Float? = null

  fun setFacing(direction: Direction?): Targetable {
    this.direction = direction
    return this
  }

  val offsetBlockPos: BlockPos
    get() {
      if (direction != null) return targetBlockPos.relative(direction!!.opposite)
      return targetBlockPos
    }

  fun simpleCopy(): Targetable {
    val out = Targetable(level, targetPos)
    out.setFacing(facing)
    return out
  }
}
